src/models/TileCoordinate.ts

```diff
@@ -0,0 +1,42 @@
+export class TileCoordinate {
+    layer: number;
+    x: number;
+    y: number;
+
+    constructor(x: number, y: number, layer: number) {
+        this.x = x;
+        this.y = y;
+        this.layer = layer;
+    }
+
+    public encode(): number {
+        return TileCoordinate.EncodeCoordinate(this);
+    }
+
+    public static EncodeCoordinate(coordinate: { x: number, y: number, layer: number }): number {
+        if (!coordinate) {
+            return -1;
+        }
+        return TileCoordinate.Encode(coordinate.x, coordinate.y, coordinate.layer);
+    }
+
+    public static Encode(x: number, y: number, layer: number): number {
+        const layerWidth = 1 << layer;
+        if (x < 0 || y < 0 || layer < 0 || layer > 12 || x >= layerWidth || y >= layerWidth) {
+            return -1;
+        }
+
+        return ((layer << 24) | (y << 12) | x);
+    }
+
+    public static Decode(encodedCoordinate: number): TileCoordinate {
```

Comment:
is it possible to use some constant variables to replace the numbers for readability? (guess this is for decoding the protobuf message)

Reply:
could do, although I'm not sure if it will help with readability. To be consistent, we'd need `const` variables for 12, 24, 19, 7, 4096 and 128, which would make the code quite a mess. 

This is for encoding and decoding a 3-component value (x, y, layer) into a single 32-bit number, in order to (a) reduce data transfer costs and (b) improve hashing behaviour. This is described in the ICD, but perhaps some comments cuold be added to the code

Reply:
hmm comments sound good, Angus thanks!